contrib/make_training_pairs.py

```diff
@@ -1,93 +1,154 @@
+        if edge.judgement == Judgement.NO_JUDGEMENT:
+            continue
+
+        judgement = edge.judgement
+        if judgement == Judgement.UNSURE:
+            judgement = Judgement.NEGATIVE
+
+        source = lazy_resolve(view, chrono_resolver, edge.source)
+        target = lazy_resolve(view, chrono_resolver, edge.target)
+        if source is None or target is None:
+            something_missing_count += 1
+            continue
+
+        if judgement == Judgement.POSITIVE:
+            positive_count += 1
+        else:
+            negative_count += 1
+
+        yield source, target, judgement
 
     log.info(
-        "Computed %d potential pairs..." % len(pairs),
-        positive=judgements.get(Judgement.POSITIVE, 0),
-        negative=judgements.get(Judgement.NEGATIVE, 0),
-        unsure=judgements.get(Judgement.UNSURE, 0),
+        f"Positive: {positive_count}, Negative: {negative_count}, Missing: {something_missing_count}"
     )
-    store = get_store(scope, external=True)
-    for idx, ((left, right), judgement) in enumerate(pairs.items()):
-        if idx > 0 and idx % 10000 == 0:
-            log.info("Exported %d pairs..." % idx)
-        left_entity = get_partial(resolver, store, left)
-        right_entity = get_partial(resolver, store, right)
-        if left_entity is None or right_entity is None:
-            continue
 
-        yield {
-            "left": left_entity.to_dict(),
-            "right": right_entity.to_dict(),
-            "judgement": judgement,
-        }
+
+def generate_groups(
+    resolver_path: Path, dataset: Dataset
+) -> Generator[Dict[str, Entity | str | int], None, None]:
+    """
+    Each list of pairs is a disconnected subgraph.
+    Negative judgements are considered edges too.
+    """
+
+    log.info("Loading resolver...")
+    complete_resolver = Resolver.load(resolver_path)
+
+    log.info("Generating pairs...")
+    g = nx.Graph()
+    for i, (source, target, judgement) in enumerate(
+        generate_pairs(resolver_path, dataset)
+    ):
+        if i % 10000 == 0 and i > 0:
+            log.info(f"Generated {i} pairs...")
+
+        source_canonical = complete_resolver.get_canonical(source.id)
+        target_canonical = complete_resolver.get_canonical(target.id)
+        # add to graph
+        g.add_node(source_canonical)
+        g.add_node(target_canonical)
+        g.add_edge(
+            source_canonical,
+            target_canonical,
+            source=source,
+            target=target,
+            judgement=judgement,
+        )
+
+    log.info("Generating disjoint subgraphs...")
+    for ix, subgraph_nodes in enumerate(nx.connected_components(g)):
+        for _, _, data in g.subgraph(subgraph_nodes).copy().edges(data=True):
+            yield {
+                "left": data["source"].to_dict(),
+                "right": data["target"].to_dict(),
+                "judgement": data["judgement"].value,
+                "group": ix,
```

Comment:
uh, fascinating, what's happening here? I'm a fan of loading this into NetworkX :)